### ISSUE TITLE
`rpcdaemon` cannot utilize all CPU cores (I/O is not the bottleneck)

### ISSUE THREAD

**Issue by zzh1996** [ledgerwatch/erigon#2245]
#### System information

Erigon version: 2021.06.5-alpha

OS & Version: Linux

Commit hash: cc360ec65

#### Expected behaviour

`rpcdaemon` should utilize all CPU cores (if I/O is not the bottleneck) when there are enough incoming RPC calls.

I have tested Go-Ethereum (geth) and it worked as expected.

#### Actual behaviour

`rpcdaemon` utilizes at most 800% ~ 900% of CPU on a 16C/32T Linux machine.

I/O percentage is less than 1% (as shown in `iotop`).

Running more RPC clients (Python scripts in my case) won't help. Each RPC call becomes slower, while `rpcdaemon` refuses to use more CPU.

Running more `rpcdaemon` instances won't help. If I run 4 instances, each of them use 200% ~ 300% CPU.

Running `rpcdaemon` without `--private.api.addr` won't help. Performance is not limited by Erigon private API.

The state of the `rpcdaemon` process is `S` in `htop` most of the time.

#### Steps to reproduce the behaviour

1. Run `rpcdaemon`.
2. Make as many RPC calls as you can. You can do this by running multiple instances of the following Python script.
3. Watch your CPU usage in `htop` and find that CPUs are not fully utilized.
4. Run multiple instances of the following script against Go-Ethereum (geth) and find that all CPUs can be utilized.

```python
import requests

if __name__ == '__main__':
    sess = requests.session()
    while True:
        sess.post('http://127.0.0.1:8545', json=[{
            'method': 'eth_call',
            'params': [
                {
                    'from': '0xbe0eb53f46cd790cd13851d5eff43d12404d33e8',
                    'to': '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
                    'gas': hex(1000000),
                    'value': hex(10 ** 18),
                    'data': '0xd0e30db0',
                },
                'latest',
            ],
            'id': i,
            'jsonrpc': '2.0',
        } for i in range(10000)])
```

**Comment by AskAlexSharov** [comment]
@zzh1996 , hi. Thank you for testing it. 
Can you please compare throughput or latency of RpcDaemon vs Geth? 

You are bottlenecked by max amount of read transactions:
`const ReadersLimit = 2000 // MDBX_READERS_LIMIT on 64bit system`

Another know hardcoded limiter is for batch requests: 
`const MaxGoroutinesPerBatchRequest = 50`

You see 1% IO because Erigon's db doesn't touch Disk if your Hot dataset fit in RAM (OS does manage eviction of cold parts of db from from it's PageCache). It means if you will call for many historical blocks - you will see more IO because all blocks do not fit in RAM (doesn't matter what data is it and how old is it - only important how Big it's Hot part).

But i wonder if in reality (under real users traffic): 
- executing more than 2000 parallel RPC requests (in same time) will only reduce overall throughput and latency? (because you will spend more time for context switching between goroutines). 
- Of course if RPC daemon run on another machine - network IO will take it's place and it make more sense to process more parallel requests in same time(because goroutines will more wait on IO). 

My opinion: having system without any boundaries is bad idea (because then 100% of clients will suffer when node overload) - but if here we can increase throughput or any other useful metric (not just make CPU busy) by increasing limits we must do it.

**Comment by zzh1996** [comment]
@AskAlexSharov
> Can you please compare throughput or latency of RpcDaemon vs Geth?

I cannot compare them fairly at this moment (but can do this later), because my Erigon and Geth node are running on seperate machines with different configuration.

> * executing more than 2000 parallel RPC requests (in same time) will only reduce overall throughput and latency? (because you will spend more time for context switching between goroutines).

I did not experiment with different settings. As I can see, if there are 32 clients sending batched requests simultaneously, there will be more than 1500 `rpcdaemon` Linux threads running, which I think is a bad idea anyway.

> My opinion: having system without any boundaries is bad idea (because then 100% of clients will suffer when node overload) - but if here we can increase throughput or any other useful metric (not just make CPU busy) by increasing limits we must do it.

There are two kinds of scenarios where we need Ethereum RPC:

1. For DApps and wallets, just like Infura, serving unspecific people.
2. For programs such as Ethereum analyzing tools and arbitrage bots, all RPC requests are under our control.

For the first scenario, I think you are right. We care much about user experience and we need boundaries to make node not overloaded. (But we can still increase throughput by making CPU busy.)

However, for the second scenario, there is no reason to limit the performance. Requests are coming in batches and are synchronous. No clients (real users) will suffer from node overload. We just want all the requests to be processed as soon as possible. I think at least we should have some options to enable this scenario.

**Comment by AskAlexSharov** [comment]
Thanks for info. 

Threads - in Golang we have goroutines which are 100x cheaper then threads. Here we are fine. 

We for sure can make limit of “threads per batch”  configurable. 

But I need few days to check if we can increase limit of read transactions.

**Comment by zzh1996** [comment]
@AskAlexSharov

I know that goroutines are cheaper than threads.

However, there's something buggy or at least not working as expected in Erigon. When there are 32 clients sending batched requests simultaneously, there are more than 1500 `rpcdaemon` Linux threads running. Yes, it's OS threads.

```
$ ps auxH | grep rpcdaemon | wc -l
1517
```

Maybe they are waiting for syscall or something else. I'm not an expert in Golang, but I don't think there should be so many OS threads even if there are more goroutines.

The same 32 clients against Geth:

```
$ ps auxH | grep geth | wc -l
37
```

**Comment by AskAlexSharov** [comment]
Yes, i think this what happen when run rpcdaemon remote: go limit amount of working threads to CPU number and can spawn unlimited amount of threads for Network IO.  ( we have option to use limited amount of network IO threads, but i found that without limit it works better ). 

Do you plan run rpcdaemon on another machine from Erigon?
If no, pass —datadir option to rpcdaemon (same as of erigon). 

I will stress-test tomorrow.

**Comment by zzh1996** [comment]
@AskAlexSharov 

I have never run rpcdaemon remotely. Erigon, rpcdaemon and clients are always on the same machine (run directly, no Docker) and I was running rpcdaemon with `--datadir` for all the results above.

**Comment by zzh1996** [comment]
@AskAlexSharov 
Are there any workarounds for now? I tried to increase `ReadersLimit` and some errors appeared. It seems that this limit is caused by some page size limit. I also tried to increase/decrease `MaxGoroutinesPerBatchRequest` but nothing changes.

**Comment by AndreaLanfranchi** [comment]
> @AskAlexSharov
> 
> I have never run rpcdaemon remotely. Erigon, rpcdaemon and clients are always on the same machine (run directly, no Docker) and I was running rpcdaemon with `--datadir` for all the results above.

"Remote" is actually a misnomer we'll try to amend.
By Remote we don't mean different processes are hosted on different machines rather the comunication amongst core and rpcdaemon are carried out exclusively by TCP sockets. The other mode is "shared" where rpcdaemon literally shares the access to the local db. In this latter mode unfortunately not all rpc requests can be satisfied (some info are not persisted into db eg. tx pool) hence the "remote" mode (or eventually the mixed one) is actually the only way to have rpcdaemon respond to all requests.

**Comment by AskAlexSharov** [comment]
We will increase max read transactions limit - in near future. waiting to support from DB side.

**Comment by AskAlexSharov** [comment]
@zzh1996 , hi. Can you try https://github.com/ledgerwatch/erigon/pull/2366
?

**Comment by zzh1996** [comment]
> @zzh1996 , hi. Can you try #2366
> ?

@AskAlexSharov I tried 04465e0, but there isn't any noticeable difference. CPU utilization, RPC throughput, number of OS threads are all the same as above.

**Comment by AskAlexSharov** [comment]
@zzh1996 , are you sure you re-build rpcdaemon and erigon? and restarted. 
I see 600% CPU utilization by your python script (on 8 core server)
And your script became bottleneck - when I added one more python script in parallel - I see 800%

**Comment by zzh1996** [comment]
> @zzh1996 , are you sure you re-build rpcdaemon and erigon? and restarted.
> I see 600% CPU utilization by your python script (on 8 core server)
> And your script became bottleneck - when I added one more python script in parallel - I see 800%

@AskAlexSharov I'm sure. I stopped and deleted everything, then run git clone and build.

**Comment by AskAlexSharov** [comment]
with `--datadir` option? 

then gimme some logs
or results of adding --pprof (and then run `go tool pprof -http=: http://localhost:6060/debug/pprof/profile\?seconds\=20`

**Comment by AskAlexSharov** [comment]
Closing, because can’t reproduce anymore. Feel free to re-open.

**Comment by zzh1996** [comment]
> with `--datadir` option?
> 
> then gimme some logs
> or results of adding --pprof (and then run `go tool pprof -http=: http://localhost:6060/debug/pprof/profile\?seconds\=20`

[pprof.rpcdaemon.samples.cpu.001.pb.gz](https://github.com/ledgerwatch/erigon/files/6934110/pprof.rpcdaemon.samples.cpu.001.pb.gz)

Version: v2021.07.05
16 Python scripts running. Each CPU% < 3%.
`rpcdaemon` CPU% at ~600% on the 16C/32T Linux machine.
IO% < 1%

logs
```
$ ./rpcdaemon --datadir /mnt/ssd/blockchain/erigon/ --http.port 8546 --private.api.addr=localhost:9090 --http.api eth,erigon,trace,debug,web3,net --pprof
INFO [08-05|03:54:55.389] Starting pprof server                    cpu="go tool pprof -lines -http=: http://127.0.0.1:6060/debug/pprof/profile?seconds=20" heap="go tool pprof -lines -http=: http://127.0.0.1:6060/debug/pprof/heap"
INFO [08-05|03:54:55.392] DB schemas compatible                    reader=3.0.0 database=3.0.0
INFO [08-05|03:54:55.392] rpc filters: subscribing to Erigon events
INFO [08-05|03:54:55.393] HTTP endpoint opened                     url=localhost:8546 ws=false ws.compression=false
INFO [08-05|03:54:55.396] interfaces compatible                    remote_db=localhost:9090 client=3.0.0 server=3.0.0
INFO [08-05|03:54:55.396] interfaces compatible                    remote_service=eth_backend client=2.1.0 server=2.1.0
INFO [08-05|03:54:55.396] interfaces compatible                    remote_service=mining      client=1.0.0 server=1.0.0
INFO [08-05|03:54:55.397] interfaces compatible                    remote_service=tx_pool     client=1.0.0 server=1.0.0
```

@AskAlexSharov 
Please re-open this issue.

**Comment by AskAlexSharov** [comment]
Hi. We fixed it in Alpha releases. I found why RPCDaemon did create much threads.

**Comment by zzh1996** [comment]
> Hi. We fixed it in Alpha releases. I found why RPCDaemon did create much threads.

Sounds great. So what's the reason and could you please point out the commit which fixes it?

**Comment by AskAlexSharov** [comment]
mdbx allow open db from multiple processes. 
to coordinate processes used mdbx.lock file
when you start new read transaction it does register it's ID in mdbx.lock
it happening inside C code
when inside C code happening IO/fsync, goland goes move current thread out out main thread-pool (where it running all goroutines) and creating new thread for thread-pool.
if try to open too much new read transactions in same time it will cause threads amount spike
I did limit amount of parallel read transactions to number of cpu's 
it increase throughput of rpcdaemon and amount of threads
by: https://github.com/ledgerwatch/erigon-lib/pull/352

if you will find new corner cases, feel free to report

**Comment by mandrigin** [comment]
I assume that is fixed, especially with 2022.07.02, feel free to reopen if there are issues
